Streaming 2i Query Performance Penalty in 1.4

engelsanchez opened the pull request:
In 1.4.2, the streaming merge sort which enables pagination creates a large performance hit to queries that might not require it (streaming). 

Making the sort optional on non-paginated 2i queries would provide greatly increased performance on queries that don't require deterministic ordering.

This issue is particularly visible when returning multi-million key results. 1-10million

seancribbs commented:
Post your graph, @angrycub! :smile:

reiddraper commented:
If possible, I'd love to see us make the sorted 2i faster, rather than adding yet another option to Riak.

Things that can be looked at:
- the merge sort code itself
- the buffer size we keep from each vnode while sorting

and probably a few others that @russelldb might have in mind.

angrycub commented:
I don't disagree that faster sorting is an excellent goal; however, I can't see a way that a sorted code path could be as fast as the unsorted code path.  Since the result has no dependence on deterministic ordering, what is the gain in paying to have the answer sorted.

angrycub commented:
![screen shot 2013-09-25 at 9 30 13 am](https://f.cloud.github.com/assets/464492/1230029/e64e388c-27c7-11e3-8c9d-f1d238f3e969.png)

This is done on a single node using the eleveldb from 1.4.2 in all of the environments and the v0 object format

angrycub commented:
Talked with @russelldb today, just capturing some of the thoughts here for posterity:
- since Riak 1.4.2 has shipped with deterministically ordered results from 2i queries, if we provided a means to bypass the sort, it would be best if you had to "opt-out" of the sort, rather than opt in.
- since all of the behavior is contained in the index FSM, we could bring back the old one with a new name and use the old one in a case where deterministic ordering is not necessary

engelsanchez commented:
Actually, only one version of Riak has shipped with the deterministic order. If we make it opt-out, it's going to be a surprise for users upgrading from all other previous versions of Riak. There is no way to avoid the confusion for some people.  The only users who would have to make a change to their code would be the ones that started or switched to 1.4.2, and made their applications rely on the order. That should be a smaller number than the vast majority of people that would get the performance penalty from pre 1.4.2 and would now have to change code to _not_ get the penalty when 2.0 comes out.

russelldb commented:
Yeah, good point.

angrycub commented:
:+1:  @engelsanchez.

reiddraper commented:
Please no, sorting has been something people have been asking for for a long time, and _should've_ been the default behavior from day one. Let's make it opt-out. IMO the use-cases for wanting unsorted results are less common, people have just had to deal with it because they had no other choice.

reiddraper commented:
As for 'surprisingly behavior', changing things _twice_ is the most surprising:

pre 1.4: unsorted
1.4: sorted default
2.0: unsorted default, sorted optional

bhigginssymc commented:
Just want to add my $0.02 for the "unsorted" option. We run nightly 2i queries that fetch 10-30 million keys. They are now taking much longer due to the sort added in 1.4. It used to take 2 - 3 minutes to retrieve and count 30 million keys using non-streamed get_index() in 1.3.2. Now it takes ~10 minutes using streamed 2i query in 1.4.2.

As for the default, I guess we don't really care, as long as the sort opt-out is simple (like a function parameter).

Cheers,
-Brian
Symantec

engelsanchez committed:
Make 2i sort optional

This addresses the performance penalty from the new default of sorting
2i results since pagination was added. When paginating, (max_results
set) sorting is always enabled. But when not, it is up to the caller.

reiddraper commented:
@engelsanchez the default is sorting _false_ now?

engelsanchez committed:
Make 2i sort default configurable

This makes the default be specified in only one place and gives people
with migration problems a way to change it without modifying their code
which has been a point of contention with this sorting business.

Cleaned up inconsistent use of sort/order. Let's go with sort flag for
now.

engelsanchez committed:
Remove use of feature branch in rebar.config

It wasn't really helping set up for testing: the deps in riak are too
complex. Maybe someday we'll get Kelly's rebar plugin for realz

engelsanchez commented:
@reiddraper I've done a round of cleanup. The default should now only come from cuttlefish.  Since either default will  surprise some customers, I've gone ahead and made it configurable. That way the worst case (people surprised and unable to update their software quickly) has a workaround.

reiddraper commented:
We've decided to defer the 'default' decision a bit. It'll get merged in with 'false' as the default for now, and will just focus on review.

engelsanchez committed:
Fix comment about 2i sort not documented

engelsanchez commented:
The comment about sort order not documented has been tweaked to remove that bit as suggested by Reid offline. This comment will likely be subject to change when we have the default value discussion.

reiddraper commented on src/riak_kv_index_fsm.erl:
```diff
@@ -87,16 +88,29 @@ init(From={_, _, _}, [Bucket, ItemFilter, Query, Timeout]) ->
+        {true, _} ->
+            true;
+        {_, undefined} ->
+            app_helper:get_env(riak_kv, secondary_index_sort_default);
```
This is kind of pedantic, but I'd prefer us to grab the default out in `riak_client`, and not have to worry about threading an `undefined` very far. Up to you on this one.

engelsanchez replied:
I'm not feeling strongly about this one, but it's not hard to imagine the FSM getting started by something other than the riak client in the future. See how riak_kv_index_pipe skips the client altogether to do its 2i query. If you are going to pick one and one only place to define a default, the single place where it is actually used seems more pure to me.

reiddraper commented on src/riak_kv_wm_index.erl:
```diff
@@ -153,18 +157,33 @@ malformed_request(RD, Ctx) ->
     ReturnTerms = normalize_boolean(string:to_lower(ReturnTerms0)),
+    Sort0 = wrq:get_qs_value(?Q_2I_SORT, RD),
+    Sort = case Sort0 of
+        undefined -> undefined;
```
re: my earlier comment, it might even be better to grab the default even further out (further from riak_client), and make riak_client require a boolean. We'd grab the default in webmachine then.

engelsanchez replied:
I think you and I have very different interpretations of DRY :). Then do the same in PB and have to maintain both paths? And remember to add it to any new path we create later? Yes, we could centralize it in some method in riak_index and make sure that every entry point calls the same code. It just seems like strange dance to me.

reiddraper replied:
> I think you and I have very different interpretations of DRY :). 

Nope, I just hadn't considered it being called from webmachine and pb. What's the argument for not putting the 'sorted' as part of the `riak_kv_index` record?

> Yes, we could centralize it in some method in riak_index and make sure that every entry point calls the same code.

Isn't that just what an API is? I'm imagining we'd just create an extra arity riak_index:to_index_query, and have the others grab the default from the app config. As I mentioned before, none of this is a big deal, but I'm enjoying the design discussion :).

engelsanchez replied:
The problem is that it's all leaky right now. to_index_query is concerned with the internal query sent to each vnode, mostly (range parameters). The other parameters are scattered (bucket, max results, return terms, etc). Ideally, we'd have this thin translation of HTTP/PB messages to this single '2i query' message to our internal API. That would check syntax at that layer, leaving the internal API to check semantics, fill in defaults, etc. So our internal API would cleanly be this single point of reference we want. Heck, we could even have a dumb transformer to avoid all this HTTP/PB code duplication... Might be worth opening an issue to looking into a through cleaning of this.

reiddraper replied:
Cool, I'm in complete agreement with that. Let's leave it as a separate issue for now, as you suggest.

engelsanchez committed:
Revert "Remove use of feature branch in rebar.config"

This reverts commit 39ac4d5a02970cc7efe6e9831e513478b9fa0c27.

reiddraper commented:
+1, but would you mind filing an issue to write a simple riak-test?

engelsanchez commented:
@reiddraper I'm open to suggestions about the riak test: I haven't figured out a way to verify this in a proper way. If sorted is not specified, results _shouldn't_ be sorted, but heck, they might be.

reiddraper commented:
> If sorted is not specified, results shouldn't be sorted, but heck, they might be.

Yeah, I wondered about that too. We can always start off by only testing that overriding the default works, which is that adding sort=true _does_ sort the results. Better than no test.

engelsanchez commented:
Agreed. I'm preparing that part of the test now. No reason to wait for later. It would also be worth to add a check that if the default is overriden via config, it sticks (again, only on the positive side unless someone comes up with something clever).

reiddraper commented:
Cool, sounds like a solid plan to me. Thanks.

engelsanchez commented:
@reiddraper did you realize that the order in which objects are returned is not by key, but by (term, key)? Somehow I wasn't thinking too much about that, but of course that is what's needed for pagination. This default order is really not that useful unless you use return_terms=true. I thought that might help in the default value discussion. But maybe this was clear in your mind already.

engelsanchez commented:
The secondary index tests were spiced up to exercise this: https://github.com/basho/riak_test/pull/444

reiddraper commented:
> did you realize that the order in which objects are returned is not by key, but by (term, key)? Somehow I wasn't thinking too much about that, but of course that is what's needed for pagination. This default order is really not that useful unless you use return_terms=true. I thought that might help in the default value discussion. But maybe this was clear in your mind already.

Yeah, though my bias has definitely been in thinking about the `$key` index. After reflecting on our discussion earlier, I I'm on board with making `false` the default for non-paginated requests. Let's just be clear on our documentation/messaging.

engelsanchez commented:
I know this is kind of closed, but to your point about messaging and documentation, I'm worried that this "sort" option is non-intuitive for all non-key/bucket, non-return_terms queries.  I suppose that as long as is documented as "sort by pagination order, that is by matching 2i term first, then object key" or something it should be OK.  If you have others thoughts about clarifying this in the code, I'm open to suggestions. Maybe pagination_sort is a better parameter name? That would force you to go "wat?" and check out the documentation to understand what it means, vs. seeing "sort" and assuming, like even I did momentarily, that it was key sort.   Coding is easy. Coding things that make sense to users is hard.

angrycub commented:
Since the pagination syntax has to sort on {term, index}, I think I would rather have a paginated sort default `return-terms` to true.  That way the most naive case makes sense when the data is returned -- otherwise the users are going to complain that their data isn't sorted, rather than looking at the docs.  If a more seasoned user wanted to disable terms, then we could allow them to do that with an explicit `return-terms = false` 

This might become it's own issue though, since this is a full mitigation of the performance issue.

reiddraper commented:
You guys are making good points, my biggest concern with changing the default to include the terms is that, I'm guessing, it will change the data-structure returned to them in the client. If this isn't the case, I'd be +1 on making it the default, but I don't want to break anyone's existing code.

reiddraper commented:
+1

slfritchie commented:
Pulling PRs out of the freezer ... wow, comment EOVERFLOW.  What remains for this ticket?

engelsanchez commented:
Development on this was backported and released in the 1.4 series, and the result of the backported to 2.0 here: https://github.com/basho/riak_kv/pull/776. 

Closing
